File: system-network/src/main/java/com/levnovikov/core_network/UrlConnectionCallExecutor.java
package com.levnovikov.core_network;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Charsets;
import com.levnovikov.core_network.request.Request;
import com.levnovikov.core_network.response.Response;
import com.levnovikov.core_network.response.ResponseBody;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;

/**
 * Created by lev.novikov
 * Date: 27/3/18.
 */

public class UrlConnectionCallExecutor implements CallExecutor {

    @Override
    public Response execute(Request request) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) request.getUrl().openConnection();
        connection.setRequestMethod(request.getMethod());
        if (request.getBody().isNotEmpty()) {
            connection.setDoOutput(true);
            OutputStream outputStream = connection.getOutputStream();
            outputStream.write(request.getBody().getContent().getBytes());
        }
        connection.connect();
        return getResponse(connection);
    }

    @VisibleForTesting
    Response getResponse(HttpURLConnection connection) throws IOException {
        return new Response.Builder()
                .code(connection.getResponseCode())
                .body(getResponseBody(connection)).build();
    }

    @VisibleForTesting
    ResponseBody getResponseBody(HttpURLConnection connection) {
        InputStream stream;
        try {
            stream = connection.getInputStream();
        } catch (IOException e) {
            stream = connection.getErrorStream();
        }

        String contentType = connection.getContentType();
        String encoding = Charsets.UTF_8.toString();
        String mediaType = "";
        if (contentType != null) {
             encoding = getCharset(contentType);
             mediaType = getMediaType(contentType);
        }
        return new ResponseBody(stream, encoding, mediaType);
    }

    private String getCharset(String contentType) {
        //TODO not implemented
        return Charsets.UTF_8.toString();
    }

    private String getMediaType(String contentType) {
        //TODO not implemented
        return "text/html";
    }
}

File: system-http-client/src/main/java/com/levnovikov/core_network/UrlConnectionCallExecutor.kt
package com.levnovikov.core_network
import com.levnovikov.core_network.request.Request
import com.levnovikov.core_network.response.Response
import com.levnovikov.core_network.response.ResponseBody

import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection

/**
 * Created by lev.novikov
 * Date: 27/3/18.
 */

class UrlConnectionCallExecutor : CallExecutor {

    @Throws(IOException::class)
    override fun execute(request: Request): Response {
        val connection = request.getUrl().openConnection() as HttpURLConnection
        connection.requestMethod = request.getMethod()
        if (request.body.isNotEmpty) {
            connection.doOutput = true
            val outputStream = connection.outputStream
            outputStream.write(request.body.content.toByteArray())
        }
        connection.connect()
        return getResponse(connection)
    }

    @Throws(IOException::class)
    private fun getResponse(connection: HttpURLConnection): Response {
        return Response.Builder()
                .code(connection.responseCode)
                .body(getResponseBody(connection)).build()
    }

    private fun getResponseBody(connection: HttpURLConnection): ResponseBody {
        var stream: InputStream? = null
        try {
            stream = try {
                connection.inputStream
            } catch (e: IOException) {
                connection.errorStream
            }

            val contentType = connection.contentType
            var encoding = "UTF-8"
            var mediaType = "text/html"
            if (contentType != null) {
                encoding = getCharset(contentType)
                mediaType = getMediaType(contentType)
            }
            return ResponseBody(stream, encoding, mediaType)
        } catch (e: Exception) {
            stream?.close()
            throw e
        }
    }

    private fun getCharset(contentType: String): String {
        //TODO not implemented
        return "UTF-8"
    }

    private fun getMediaType(contentType: String): String {
        //TODO not implemented
        return "text/html"
    }
}
